ros2_reduct_agent/recorder.py

```diff
@@ -0,0 +1,376 @@
+import asyncio
+import importlib
+from collections import defaultdict
+from tempfile import SpooledTemporaryFile
+from typing import Any
+
+import rclpy
+from mcap_ros2.writer import Writer as McapWriter
+from rclpy.node import Node
+from rclpy.qos import QoSProfile
+from rclpy.subscription import Subscription
+from rclpy.time import Time
+from reduct import Client
+from rosbag2_py import LocalMessageDefinitionSource
+
+from .config_models import FilenameMode, PipelineConfig, PipelineState, StorageConfig
+
+
+class Recorder(Node):
+    """ROS2 node that records selected topics to ReductStore."""
+
+    def __init__(self, **kwargs):
+        super().__init__(
+            "recorder",
+            allow_undeclared_parameters=True,
+            automatically_declare_parameters_from_overrides=True,
+            **kwargs,
+        )
+        self.logger = self.get_logger()
+        self.warned_topics: set[str] = set()
+
+        # Parameters
+        self.storage_config = self.load_storage_config()
+        self.pipeline_configs = self.load_pipeline_config()
+
+        # ReductStore
+        self.client = Client(
+            self.storage_config.url, api_token=self.storage_config.api_token
+        )
+        self.bucket = None
+        self.loop = asyncio.get_event_loop()
```

Comment:
Strange that I don't see warnings from this line but in the lates Python ersion this is the preferable way to get a loop: https://stackoverflow.com/questions/73884117/how-to-replace-asyncio-get-event-loop-to-avoid-the-deprecationwarning